### PR TITLE
Introduce a `__revert` intrinsic a disallow `rvrt`, `ret` and `retd` in `asm` blocks

### PR DESCRIPTION
**Pull request by mohammadfawaz**
Closes https://github.com/FuelLabs/sway/issues/2884

Main changes:
1. Disallowing `rvrt`, `ret`, and `retd` in `asm` blocks.
2. Remove all checks that look for the opcodes above.
3. Add `__revert` intrinsic and make its type `Unknown` for now. It will be `Never` in the future.
4. Add an IR instruction `revert` with no type. There are still some things that should be ironed out related to that.

Also closes https://github.com/FuelLabs/sway/issues/2822.

https://github.com/FuelLabs/sway/issues/2876 is not fixed yet though. This is related to the issues mentioned in (4) above.

### CODE REVIEW

File: sway-core/src/ir_generation/function.rs [hunk 1]
```diff
@@ -643,6 +643,13 @@ impl FnCompiler {
                     .ins(context)
                     .binary_op(op, lhs_value, rhs_value))
             }
+            Intrinsic::Revert => {
+                let revert_code_val =
+                    self.compile_expression(context, md_mgr, arguments[0].clone())?;
+
+                // The `revert` instruction
+                Ok(self.current_block.ins(context).revert(revert_code_val))
+            }
```
Comment (otrho) [inline sway-core/src/ir_generation/function.rs:652]: Is there a span we can attach here?

**Review by otrho — COMMENTED** [review]
Looks good other than tiny question below. :slightly_smiling_face: 

This is going to conflict with my calls branch in several places. :slightly_frowning_face:

File: sway-core/src/ir_generation/function.rs (hunk 1 above)
Comment (mohammadfawaz) [reply sway-core/src/ir_generation/function.rs]: Done!

**Comment by mohammadfawaz** [comment]
> This is going to conflict with my calls branch in several places. 🙁

Alright, so this PR is now rebased over your branch - I resolved all the conflicts. I will continue to be a draft PR until your PR is merged. It should instantly be ready to be merged right after yours go in.

**Comment by otrho** [comment]
> > This is going to conflict with my calls branch in several places. 🙁
> 
> Alright, so this PR is now rebased over your branch - I resolved all the conflicts. I will continue to be a draft PR until your PR is merged. It should instantly be ready to be merged right after yours go in.

Oh you didn't need to do that!  Thanks!  ❤️

**Review by vaivaswatha — APPROVED** [review]

**Review by vaivaswatha — APPROVED** [review]

**Review by sezna — APPROVED** [review]

**Review by emilyaherbert — APPROVED** [review]

**Pull request merged**